autogpt/memory/message_history.py

```diff
@@ -0,0 +1,186 @@
+from autogpt.log_cycle.log_cycle import PROMPT_SUMMARY_FILE_NAME, SUMMARY_FILE_NAME
+from autogpt.logs import logger
+from autogpt.memory.vector import VectorMemory
+
+
+class MessageHistory(list[Message]):
+    agent: Agent
+
+    summary: str
+
+    last_trimmed_index: int
+
+    def __init__(self, agent: Agent):
+        self.agent = agent
+        self.summary = "I was created."
+        self.last_trimmed_index = 0
+
+    def trim_messages(
+        self,
+        current_message_chain: list[Message],
+    ) -> tuple[Message, list[Message]]:
+        """
+        Returns a list of trimmed messages: messages which are in the message history
+        but not in current_message_chain.
+
+        Args:
+            current_message_chain (list[Message]): The messages currently in the context.
+
+        Returns:
+            Message: A message with the new running summary after adding the trimmed messages.
+            list[Message]: A list of messages that are in full_message_history with an index higher than last_trimmed_index and absent from current_message_chain.
+        """
+        # Select messages in full_message_history with an index higher than last_trimmed_index
+        new_messages = [
+            msg for i, msg in enumerate(self) if i > self.last_trimmed_index
+        ]
+
+        # Remove messages that are already present in current_message_chain
+        new_messages_not_in_chain = [
+            msg for msg in new_messages if msg not in current_message_chain
+        ]
+
+        # agent.history.archive(
+        #     new_messages_not_in_chain,
+        #     agent.memory,
+        # )
+
+        new_summary_message = self.update_running_summary(
+            new_events=new_messages_not_in_chain
+        )
+
+        # Find the index of the last message processed
+        if new_messages_not_in_chain:
+            last_message = new_messages_not_in_chain[-1]
+            self.last_trimmed_index = self.index(last_message)
+
+        return new_summary_message, new_messages_not_in_chain
+
+    def archive(
+        self,
+        up_to_index: int,
+        permanent_memory: VectorMemory,
+    ):
+        """Saves messages up to the given index to permanent memory"""
+
+        while up_to_index >= 0:
+            ai_message = self[up_to_index]
+            if is_string_valid_json(ai_message["content"], LLM_DEFAULT_RESPONSE_FORMAT):
+                next_message = self[up_to_index + 1]
+                memory_to_add = MemoryItem.from_ai_action(ai_message, next_message)
+                logger.debug(f"Storing the following memory: {memory_to_add}")
+                permanent_memory.add(memory_to_add)
+
+            up_to_index -= 1
+
+    def summary_message(self) -> Message:
+        return {
+            "role": "system",
+            "content": f"This reminds you of these events from your past: \n{self.summary}",
+        }
+
+    def update_running_summary(self, new_events: list[Message]) -> Message:
```

Comment:
I think it should well outside the responsibilities of the message history object to be invoking language models.

Reply:
Ignore for now